refactoringminer/src/main/java/gr/uom/java/xmi/UMLTypePsiParser.java

```diff
@@ -0,0 +1,207 @@
+                .collect(Collectors.toList());
+            return new ListCompositeType(umlTypes, ListCompositeType.Kind.UNION);
+        } else if (type instanceof PsiWildcardType) {
+            PsiWildcardType wildcardType = (PsiWildcardType) type;
+            if (wildcardType.isBounded()) {
+                PsiTypeElement bound = (PsiTypeElement) typeElement.getLastChild();
+                return new WildcardType(extractTypeObject(file, filePath, bound,
+                    wildcardType.getBound()), !wildcardType.isSuper());
+            } else {
+                return new WildcardType(null, true);
+            }
+        } else if (type instanceof PsiEllipsisType) {
+            typeElement = findFirstChildOfType(typeElement, PsiTypeElement.class);
+            UMLType ellipsisType = extractTypeObject(file, filePath, typeElement);
+            ellipsisType.arrayDimension++;
+            return ellipsisType;
+        } else if (type instanceof PsiArrayType) {
+            if (typeElement.getType() instanceof PsiArrayType) {
+                typeElement = findFirstChildOfType(typeElement, PsiTypeElement.class);
+            }
+            UMLType arrayType = extractTypeObject(file, filePath, typeElement);
+            arrayType.arrayDimension = type.getArrayDimensions();
+            return arrayType;
+        } else {
+            PsiJavaCodeReferenceElement typeReference =
+                findFirstChildOfType(typeElement, PsiJavaCodeReferenceElement.class);
+            if (typeReference != null) {
+                return extractTypeObject(file, filePath, typeReference);
+            }
+            PsiKeyword typeKeyword = findFirstChildOfType(typeElement, PsiKeyword.class);
+            return new LeafType(Formatter.format(typeKeyword));
+        }
+    }
+
+    private static List<Reference> parseReferences(PsiFile file, String filePath, PsiJavaCodeReferenceElement typeReference) {
+        PsiJavaCodeReferenceElement reference = findFirstChildOfType(typeReference, PsiJavaCodeReferenceElement.class);
+        List<Reference> preceding;
+        if (reference == null) {
+            preceding = new ArrayList<>();
+        } else {
+            preceding = parseReferences(file, filePath, reference);
+        }
+
+        String identifier = Formatter.format(findFirstChildOfType(typeReference, PsiIdentifier.class));
+
+        PsiReferenceParameterList typeParameters = findFirstChildOfType(typeReference, PsiReferenceParameterList.class);
+        List<UMLType> generics = parseGenerics(file, filePath, typeParameters);
+
+        LocationInfo locationInfo = new LocationInfo(file, filePath, typeReference, LocationInfo.CodeElementType.TYPE);
+
+        Reference referenceElement = new Reference(identifier, generics, locationInfo);
+        preceding.add(referenceElement);
+        return preceding;
+    }
+
+    private static List<UMLType> parseGenerics(PsiFile file, String filePath,
```

Comment:
what about renaming `processGenericTypes`?

Reply:
GenericTypes ok. But I think `parseGenericTypes` is more appropriate because process implies `void` return type while parse function return type is what he parse. `extract` sounds good too